lib/cli/do_nfs_volumes.js

```diff
@@ -0,0 +1,341 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2017, Joyent, Inc.
+ */
+
+/*
+ * The 'sdcadm experimental nfs-volumes' CLI subcommand.
+ *
+ * Currently, NFS shared volumes are still at the prototype stage, and
+ * they must be enabled/disabled by setting a SAPI configuration flag.
+ */
+
+var assert = require('assert-plus');
+var https = require('https');
+var once = require('once');
+var util = require('util'),
+    format = util.format;
+var vasync = require('vasync');
+
+
+var common = require('../common');
+var errors = require('../errors');
+var shared = require('../procedures/shared');
+
+function getCoreServiceImageVersion(coreServiceName, options, callback) {
+    assert.string(coreServiceName, 'coreServiceName');
+    assert.object(options, 'options');
+    assert.object(options.imgapiClient, 'options.imgapiClient');
+    assert.object(options.sapiClient, 'options.sapiClient');
+    assert.object(options.sdcApplication, 'options.sdcApplication');
+    assert.object(options.vmapiClient, 'options.vmapiClient');
+    assert.func(callback, 'callback');
+
+    var context = {};
+    var imgapiClient = options.imgapiClient;
+    var sapiClient = options.sapiClient;
+    var sdcApp = options.sdcApplication;
+    var vmapiClient = options.vmapiClient;
+
+    vasync.pipeline({arg: context, funcs: [
+        function getSvc(ctx, next) {
+            sapiClient.listServices({
+                name: coreServiceName,
+                application_uuid: sdcApp.uuid
+            }, function (svcErr, svcs) {
+                if (svcErr) {
+                    next(svcErr);
+                    return;
+                } else if (svcs && svcs.length > 0) {
+                    ctx.svc = svcs[0];
+                }
+                next();
+            });
+        },
+
+        function getInst(ctx, next) {
+            assert.object(ctx.svc, 'ctx.svc');
+
+            var filter = {
+                service_uuid: ctx.svc.uuid
+            };
+
+            sapiClient.listInstances(filter, function (err, insts) {
+                if (err) {
+                    return next(new errors.SDCClientError(err, 'sapi'));
+                } else if (insts && insts.length) {
+                    // Note this doesn't handle multiple insts.
+                    ctx.inst = insts[0];
+                    vmapiClient.getVm({
+                        uuid: ctx.inst.uuid
+                    }, function (vmErr, instVm) {
+                        if (vmErr) {
+                            return next(vmErr);
+                        }
+
+                        ctx.instVm = instVm;
+                        next();
+                    });
+                } else {
+                    next();
+                }
+            });
+        },
+        function getInstImage(ctx, next) {
+            assert.object(ctx.instVm, 'ctx.instVm');
```

Comment:
That's a good point, but instead of asserting, I think we should call next with an error in this case.

Rationale: we assert on `ctx.instVm` because otherwise we should have interrupted the vasync workflow at the previous step. If `ctx.instVm` doesn't have a valid `image_uuid`, I would consider it an operational error.

What do you think?